### PR TITLE
Fix block merging pass

### PR DESCRIPTION
**Pull request by jdahm**
There is a small bug in block merging that results in certain blocks merging that have data dependencies. This fundamentally changes the code result (see #184). This PR fixes that and adds a test for it.

To do:
- [x] Determine if this can be reproduced without the test infrastructure ability to start from multiple ComputationBlocks
- [x] Clean up the ugly test hacks!

Blocked by #198.

### CODE REVIEW

File: tests/definition_setup.py [hunk 1]
```diff
@@ -72,42 +72,64 @@ def make_assign(
...
+    fields: List[str],
+    domain: Domain,
+    info: List[Tuple[BodyType, IterationOrder, Tuple[int, int]]],
 ) -> StencilDefinition:
     api_signature = [ArgumentInfo(name=n, is_keyword=False) for n in fields]
-    tmp_fields = {i[0] for i in body}.union({i[1] for i in body}).difference(fields)
+    bodies = []
+    for definition in info:
+        bodies.extend(definition[0])
+    tmp_fields = {i[0] for i in bodies}.union({i[1] for i in bodies}).difference(fields)
```
Comment (DropD) [inline tests/definition_setup.py:85]: Are you sure this does the same thing as before?
it seems like the expression `{i[X] for i in bodies} will now create a set of `BodyType` rather than field names...

Comment (DropD) [reply tests/definition_setup.py:85]: Ah, I think I miscounted nesting levels. It has become a little hard to follow compared to before

File: tests/test_unittest/test_merge_blocks_pass.py [hunk 2]
```diff
@@ -106,6 +106,23 @@ def test_merge_write_after_read_k_parallel(
...
+        fields=["out", "in", "tmp"],
+        info=[
+            ([("tmp", "in", (0, 0, 1))], IterationOrder.FORWARD, (1, -1)),
+            ([("out", "tmp", (0, 0, -1))], IterationOrder.FORWARD, (1, -2)),
+        ],
+    )
+    transform_data = merge_blocks_pass(transform_data)
+    # not allowed to be merged into the same stage: first block should have 2 IJ blocks
+    assert len(transform_data.blocks[0].ij_blocks) == 2
+
```
Comment (DropD) [inline tests/test_unittest/test_merge_blocks_pass.py:124]: Ideally we should also check that the two multistages are kept in the right order. Especially since a reordering pass is planned.

Comment (jdahm) [reply tests/test_unittest/test_merge_blocks_pass.py:124]: Good idea. I'll make that change.

File: tests/definition_setup.py [hunk 3]
```diff
@@ -72,42 +72,64 @@ def make_assign(
...
+    name: str,
+    fields: List[str],
+    domain: Domain,
+    info: List[Tuple[BodyType, IterationOrder, Tuple[int, int]]],
 ) -> StencilDefinition:
     api_signature = [ArgumentInfo(name=n, is_keyword=False) for n in fields]
-    tmp_fields = {i[0] for i in body}.union({i[1] for i in body}).difference(fields)
+    bodies = []
+    for definition in info:
+        bodies.extend(definition[0])
```
Comment (jdahm) [inline tests/definition_setup.py:84]: (Notes for myself): We can use flatten here.

File: tests/definition_setup.py (hunk 1 above)
Comment (DropD) [reply tests/definition_setup.py:85]: I have some ideas, I will get back to you when I have tried them.

File: tests/test_unittest/test_merge_blocks_pass.py [hunk 4]
```diff
@@ -166,6 +185,24 @@ def test_merge_read_after_write_k_sequential(
...
+        fields=["out1", "out2", "in1", "in2"],
+        info=[
+            ([("out1", "in1", (0, 0, 0))], IterationOrder.PARALLEL, (0, 0)),
+            ([("out2", "in2", (0, 0, 0))], IterationOrder.PARALLEL, (1, -1)),
+        ],
+    )
+    transform_data = merge_blocks_pass(transform_data)
+    # not allowed to be merged into the same stage: overlapping intervals
+    assert len(transform_data.blocks) == 1
+    assert len(transform_data.blocks[0].ij_blocks) == 2
```
Comment (jdahm) [inline tests/test_unittest/test_merge_blocks_pass.py:203]: @eddie-c-davis does it make sense to do the same and check to make sure the order stays the same?

File: tests/definition_setup.py (hunk 3 above)
Comment (jdahm) [reply tests/definition_setup.py]: `itertools.chain.from_iterable` worked.

File: tests/test_unittest/test_merge_blocks_pass.py (hunk 4 above)
Comment (eddie-c-davis) [reply tests/test_unittest/test_merge_blocks_pass.py:203]: Sure, I will change it.

Comment (jdahm) [reply tests/test_unittest/test_merge_blocks_pass.py:203]: Nice work!

File: tests/definition_setup.py (hunk 1 above)
Comment (DropD) [reply tests/definition_setup.py:84]: @jdahm I made a pull request into this branch on your fork that generalizes this setup in a more extensible way.

If you like it, you can simply merge it, else I hope you can build on it.

File: tests/test_unittest/test_ir_maker.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def test_AugAssign():
+    ir_maker = IRMaker(None, None, None, domain=None, extra_temp_decls=None)
+    aug_assign = ast.parse("a += 1").body[0]
+
+    _, result = ir_maker.visit_AugAssign(aug_assign)
+
+    assert isinstance(result.value, BinOpExpr)
+    assert result.value.op == BinaryOperator.ADD
```
Comment (jdahm) [inline tests/test_unittest/test_ir_maker.py:14]: I think we should keep this test! It will pass once we merge with master after merging #198.

File: tests/definition_setup.py (hunk 1 above)
Comment (jdahm) [reply tests/definition_setup.py]: Merged! Thanks again.

**Comment by jdahm** [comment]
This should also be ready to go after merging with master after merging #198.

**Comment by DropD** [comment]
bors try

File: tests/test_unittest/test_ir_maker.py (hunk 5 above)
Comment (DropD) [reply tests/test_unittest/test_ir_maker.py:14]: I actually intended to contribute that one to #198, sorry for complicating things

**Comment by DropD** [comment]
The test faillure on gitlab had nothing to do with this PR, it succeeded after rerunning. This means the tests pass on the current master (the other checks were run before merging #198).

**Review by DropD — APPROVED** [review]
